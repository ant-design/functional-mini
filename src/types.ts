export enum ETargetPlatform {
  alipay = 'alipay',
  wechat = 'wechat',
}

export enum EElementType {
  page = 'page',
  component = 'component',
}
