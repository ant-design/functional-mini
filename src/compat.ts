export * from './3rd-party/compat.js';

import compat from './3rd-party/compat.js';
export default compat;
