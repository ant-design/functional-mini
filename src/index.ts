import { setLogSwitch } from './utils.js';

export { setLogSwitch };
