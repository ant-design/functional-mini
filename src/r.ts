import { preactHooks, preact, serverRender } from './3rd-party/preact.js';
import { virtualDocument } from './3rd-party/virtual-document.js';
export { act } from './3rd-party/preact-test-utils.js';

export { serverRender };

const { createContext, h, render, options } = preact;

export { preactHooks };

export type VNode = preact.VNode;

export const React = {
  createElement: h,
  h: h,
  createContext,
  render,
  ...preactHooks,
};

options.requestAnimationFrame = (cb) => {
  return setTimeout(cb);
};

options.debounceRendering = (cb) => {
  return setTimeout(cb);
};

export const mountElement = function (element: any) {
  const entryDom = virtualDocument.createElement('div') as any;
  render(element, entryDom);

  return {
    unmount() {
      render(h('div', {}, 'unmounted'), entryDom);
    },

    //@ts-expect-error
    update(newElement) {
      render(newElement, entryDom);
    },

    toString() {
      return entryDom.innerHTML;
    },

    dom: entryDom,
  };
};
