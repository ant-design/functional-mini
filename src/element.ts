import { React, VNode, act, mountElement, serverRender } from './r.js';
import {
  log as globalLog,
  error as globalError,
  EComponent2Status,
  getComponent2Status,
  updateComponent2Status,
  getIdFromAppxInstance,
  instanceKeyPropNames,
} from './utils.js';

import HandlersController from './handlers.js';
import { IElementContext, reactContext, useSyncMiniData } from './hooks.js';
import { ETargetPlatform, EElementType } from './types.js';
import { platformConfig } from './platform.js';

const FUNCTIONAL_MINI_PAGE_DOM_PLACEHOLDER = 'MINIFISH_PAGE_DOM_PLACEHOLDER';
const DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA =
  'DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA'; // 绕过所有函数方法，直接返回 data。不要在生产环境使用！

export interface IAppxOptions {
  data: Record<string, any>;
  options: Record<string, string>;
  methods?: Record<string, (args?) => void>;
  [handlerName: string]: any;
}

export type TElementFunction<TProps> = (props: TProps) => {
  [axmlKey: string]: any;
};

export interface IInstanceMap {
  [id: string]: {
    appxContext: IElementContext;
    appxId: string;
    elementFn: any;
    elementInstance?: any;
    pendingProps: any;
    unmounted: boolean;
    propKeys: string[];
  };
}

function compositeElementWithContext(
  id,
  elementFn,
  appxContext: IElementContext,
  pendingProps,
) {
  const el = React.createElement(elementFn, pendingProps);
  // eslint-disable-next-line react/no-children-prop
  return React.createElement(
    reactContext.Provider,
    { value: appxContext, key: id, children: [el] },
    el,
  ) as VNode;
}

export function flushReactTree(elementMap: IInstanceMap) {
  const children: VNode[] = [];
  for (const id in elementMap) {
    if (!Object.prototype.hasOwnProperty.call(elementMap, id)) continue;

    const item = elementMap[id];
    if (item.unmounted) {
      continue;
    }
    if (item.elementInstance && !item.pendingProps) {
      children.push(item.elementInstance);
      continue;
    }

    const contextElement = compositeElementWithContext(
      id,
      item.elementFn,
      item.appxContext,
      item.pendingProps,
    );
    children.push(contextElement);
    item.elementInstance = contextElement;
    item.pendingProps = null;
  }

  const parent = React.createElement('div', {}, children);
  return parent;
}

export function functionalMiniElement<TProps>(
  element: TElementFunction<TProps>,
  displayName = '' /* 用于问题排查，和小程序 axml 无关 */,
  elementType: EElementType,
  defaultProps?: TProps,
  _targetPlatform?: ETargetPlatform,
) {
  let targetPlatform: ETargetPlatform =
    _targetPlatform ?? ETargetPlatform.alipay;

  // 配置客户端环境
  const {
    pageEvents,
    componentEvents,
    buildOptions,
    componentLifeCycleToMount,
    componentLifeCycleToUnmount,
    pageLifeCycleToMount,
    pageLifeCycleToUnmount,
    getPropsFromInstance,
  } = platformConfig[targetPlatform];

  displayName = displayName || element.name;
  if (!displayName) {
    console.warn(
      '为了方便问题排查，请传入组件的 displayName 参数，或不要使用匿名函数组件 https://medium.com/@stevemao/do-not-use-anonymous-functions-to-construct-react-functional-components-c5408ec8f4c7',
    );
  }

  const nameTag = `[${elementType}/${displayName || '(unnamed)'}]`;
  const log = (...args) => {
    globalLog(nameTag, ...args);
  };
  const logErrorAndThrow = (err: Error) => {
    if (!err) return;
    err.message = `${nameTag} ${err.message}`;
    globalError(err);
    throw err;
  };

  if (elementType === 'component') {
    if (getComponent2Status() === EComponent2Status.UNKNOWN) {
      updateComponent2Status();
    }
    if (getComponent2Status() === EComponent2Status.INVALID) {
      // 如果还是 Unknown，就先不管了
      throw new Error(
        `无法注册 ${nameTag} 组件，因为当前小程序环境未开启 component2。配置入口在 mini.project.json : { "compileOptions": { "component2": true } }，或在 IDE 详情>项目配置中勾选`,
      );
    }
  }

  const platformExposedEvents: string[] =
    elementType === 'page' ? pageEvents : componentEvents;

  const defaultPropKeys = Object.keys(defaultProps || {});
  let observers = {};
  const elementMap: IInstanceMap = {};
  let commonTestRenderer;
  function updateReactTree() {
    act(() => {
      const parent = flushReactTree(elementMap);
      if (!commonTestRenderer) {
        commonTestRenderer = mountElement(parent);
      } else {
        commonTestRenderer.update(parent);
      }
    });
  }

  const WrappedElementFn = function (props: any) {
    let miniData: any;
    if (props && props[DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA]) {
      miniData = props[DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA];
      if (typeof miniData === 'string' && miniData.indexOf('%7B') === 0) {
        miniData = JSON.parse(decodeURIComponent(miniData));
      }
    } else {
      try {
        miniData = element.call(undefined, props);
      } catch (e) {
        e.message = `渲染出错 ${e.message}`;
        return logErrorAndThrow(e);
      }
    }
    if (typeof miniData === 'undefined') {
      log('函数组件没有返回渲染数据，请检查代码逻辑');
      miniData = {};
    } else if (typeof miniData !== 'object') {
      const e = new Error(
        `函数组件返回的渲染数据不合法，收到的类型为 ${typeof miniData}`,
      );
      return logErrorAndThrow(e);
    }
    // 不允许和 props 里的 key 重复
    const conflictKeys: string[] = [];
    const dataKeys = Object.keys(miniData) || [];
    for (const key in props) {
      if (dataKeys.indexOf(key) >= 0) {
        conflictKeys.push(key);
      }
    }
    if (conflictKeys.length > 0) {
      const e = new Error(
        `小程序自定义组件返回的渲染数据和 props 里的 key 重复：${conflictKeys.join(
          ', ',
        )}`,
      );
      return logErrorAndThrow(e);
    }

    // 上面的 early-return 都是检查和抛错，忽略 hooks 规则
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useSyncMiniData(miniData || {});
    return React.createElement('div', {}, FUNCTIONAL_MINI_PAGE_DOM_PLACEHOLDER);
  };

  // 在 onload 的时候，正式创建一个 react 组件
  function hookLoadToMountReactComponent(this: any, ...args: any[]) {
    const appxInstance = this;
    const id = getIdFromAppxInstance(appxInstance);
    if (elementMap[id])
      throw new Error(
        `duplicate id of appx instance, this might be a bug of minifish hooks. id: ${id}`,
      );
    const context = generateInstanceContext(appxInstance, false);
    appxInstance[instanceKeyPropNames] = defaultPropKeys;
    log('will mount react component');
    let initProps = {};
    // 微信和支付宝的行为是有所不同的：微信这时候还是 defaultProps 的内容，支付宝已经是真的数据了
    if (elementType === 'component') {
      initProps = getPropsFromInstance(appxInstance, defaultPropKeys);
    }

    elementMap[id] = {
      appxContext: context,
      appxId: id,
      elementFn: WrappedElementFn,
      pendingProps: initProps,
      unmounted: false,
      propKeys: defaultPropKeys,
    };
    updateReactTree();

    // 触发 onLoad
    if (elementType === 'page') {
      log('React 页面（Page）已经 Mount，开始触发 onLoad');
      return context.handlersController.callHandlers(
        pageLifeCycleToMount,
        appxInstance,
        args,
      ); // 直接调用controller 内部的方法，插队执行 onLoad
    } else {
      log(
        `React 组件（Component）已经 Mount，开始触发 ${componentLifeCycleToMount}`,
      );
      return context.handlersController.callHandlers(
        componentLifeCycleToMount,
        appxInstance,
        args,
      );
    }
  }

  function dispatchNewProps(appxInstance, nextProps) {
    const id = getIdFromAppxInstance(appxInstance);
    const instance = elementMap[id];
    instance.pendingProps = nextProps;
    updateReactTree();
  }

  function hookUnloadToUnmount(this: any) {
    const appxInstance = this;
    const id = getIdFromAppxInstance(appxInstance);
    log(`will unmount react element of ${id}`);
    if (!elementMap[id]) {
      log(
        `找不到 id 为 ${id} 的实例，请检查是否开启了 component2 。这也可能是 Minifish 的 bug。`,
      );
    }
    elementMap[id].unmounted = true;
    updateReactTree();
    // TODO: 清理 appx context
  }

  const handlersController = new HandlersController(nameTag);
  const anyUnknownContext = (ctx) => {
    if (!ctx) throw new Error('ctx is required');
    const id = getIdFromAppxInstance(ctx);
    return !elementMap[id];
  };
  const anyContext = () => true;
  if (elementType === 'page') {
    handlersController.addHandler(
      pageLifeCycleToMount,
      anyUnknownContext,
      hookLoadToMountReactComponent,
    );
    handlersController.addHandler(
      pageLifeCycleToUnmount,
      anyContext,
      hookUnloadToUnmount,
    );
  } else {
    // 组件的关键生命周期
    handlersController.addHandler(
      componentLifeCycleToMount,
      anyUnknownContext,
      hookLoadToMountReactComponent,
    );
    handlersController.addHandler(
      componentLifeCycleToUnmount,
      anyContext,
      hookUnloadToUnmount,
    );
    // 把 props 变更分发到对应的组件里
    if (targetPlatform === ETargetPlatform.alipay) {
      handlersController.addHandler(
        'deriveDataFromProps',
        anyContext,
        function hookDeriveDataFromProps(this: any, nextProps) {
          if (this.props && this.props === nextProps) return; // 可能是setData触发的，不要死循环了
          return dispatchNewProps(this, nextProps);
        },
      );
    } else if (targetPlatform === ETargetPlatform.wechat) {
      const props = defaultPropKeys.join(', ');
      log(`将注册以下 key 的 props 更新：${props}`);
      observers = {
        // 忽略函数参数，直接从 this 里面找
        // 搞不懂小程序的设计逻辑，为什么非要在 props 里夹着 data
        [props](this: any, ...args) {
          log('observer is being called', args);
          const newProps = getPropsFromInstance(this, defaultPropKeys);
          return dispatchNewProps(this, newProps);
          //
        },
        ...observers,
        // []
      };
    }
  }

  // 做一次预渲染，获取所有 appx 需要的属性
  const generateInstanceContext = (
    instance,
    ifServerRender: boolean,
  ): IElementContext => {
    return {
      instance,
      handlersController,
      ifServerRender,
      debugLog: log,
      platformConfig: platformConfig[targetPlatform],
    };
  };

  // 收集 initData
  let initData = {};
  const fakeAppxInstance = {
    $id: `_minifish_hooks_pre_render_${Math.random()}`,
    setData(data) {
      initData = {
        ...(initData || {}),
        ...data,
      };
    },
  };
  const fakeCtx = generateInstanceContext(fakeAppxInstance, true);
  const serverEl = compositeElementWithContext(
    fakeAppxInstance.$id,
    WrappedElementFn,
    fakeCtx,
    defaultProps || {},
  );
  serverRender(serverEl);
  log('serverRendered with initData', initData);

  // 拼装对象，喂给 appx。产出的配置应该只能喂一次，喂多了 appx 不认。
  handlersController.lockHandlerNames();

  // 把生命周期和一般事件处理分开归类，因为组件的事件处理要多裹一层
  const lifeCycleHandlers = {};
  const userEventHandlers = {};
  const handlers = handlersController.getHandlersImplProxy();
  for (const name in handlers) {
    if (platformExposedEvents.indexOf(name) >= 0) {
      lifeCycleHandlers[name] = handlers[name];
    } else {
      userEventHandlers[name] = handlers[name];
    }
  }

  const finalOptions = buildOptions(
    elementType,
    defaultProps,
    initData,
    lifeCycleHandlers,
    userEventHandlers,
    null,
    observers,
  );
  log('element options', finalOptions);
  return finalOptions;
}

export function tellIfFunctionalMiniContent(elementJson) {
  return elementJson?.children[0] === FUNCTIONAL_MINI_PAGE_DOM_PLACEHOLDER;
}
