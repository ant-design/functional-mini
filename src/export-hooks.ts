import { React } from './r.js';
import { useEvent } from './hooks.js';

const { useState, useEffect, useMemo, useCallback } = React;

export { useCallback, useEffect, useMemo, useState };

export { useEvent };
