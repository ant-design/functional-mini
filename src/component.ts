import { functionalMiniElement, TElementFunction } from './element.js';
import { getLifeCycleHooks } from './hooks.js';
import { alipayComponentEvents, commonComponentEvents } from './platform.js';
import { EElementType, ETargetPlatform } from './types.js';

export function alipayComponent<TProps extends Record<string, any>>(
  element: TElementFunction<TProps>,
  defaultProps?: TProps,
) {
  return functionalMiniElement(
    element,
    '',
    EElementType.component,
    defaultProps,
    ETargetPlatform.alipay,
  );
}

export function wechatComponent<TProps extends Record<string, any>>(
  element: TElementFunction<TProps>,
  defaultProps?: TProps,
) {
  return functionalMiniElement(
    element,
    '',
    EElementType.component,
    defaultProps,
    ETargetPlatform.wechat,
  );
}

export const useAttached = getLifeCycleHooks(commonComponentEvents.attached);
export const useReady = getLifeCycleHooks(commonComponentEvents.ready);
export const useMoved = getLifeCycleHooks(commonComponentEvents.moved);
export const useDetached = getLifeCycleHooks(commonComponentEvents.detached);

export const useDidMount = getLifeCycleHooks(
  alipayComponentEvents.didMount,
  undefined,
  ETargetPlatform.alipay,
);

// onInit, deriveDataFromProps, didUnmount, didUpdate 是不开放 Hooks 的，请使用 React 生命周期方法

export * from './export-hooks.js';

export { useComponent } from './hooks.js';
