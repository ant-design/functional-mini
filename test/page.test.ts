/* eslint-disable no-prototype-builtins */
import {
  assert,
  describe,
  expect,
  test,
  vi,
  beforeEach,
  afterEach,
} from 'vitest';
import { React } from '../src/r';
import { mountAlipayPage, setupAlipayEnv, setupWechatEnv } from './common';
import {
  alipayPage,
  useOnShareAppMessage,
  useOnTitleClick,
  wechatPage,
  useEvent,
  useOnLoad,
  useOnShow,
} from '../src/page';
import { TElementFunction, functionalMiniElement } from '../src/element';
import { EElementType, ETargetPlatform } from '../src/types';

function functionalPage<TProps>(
  element: TElementFunction<TProps>,
  displayName = '',
  targetPlatform?: ETargetPlatform,
) {
  return functionalMiniElement(
    element,
    displayName,
    EElementType.page,
    {} as TProps,
    targetPlatform,
  );
}

const { useEffect, useState, useMemo } = React;

const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

describe('page - common and alipay', () => {
  let off: () => void;
  beforeEach(() => {
    off = setupAlipayEnv();
  });

  afterEach(() => {
    off();
  });

  test('minium get appx options', async () => {
    const fn = vi.fn();
    const fnOff = vi.fn();
    const fnLifeCycle = vi.fn();
    const C = function () {
      useOnLoad(fnLifeCycle, []);
      const [d] = useState('d-initValue');
      useEffect(fn, []);
      useEffect(() => fnOff, []);
      return {
        foo: 'aa',
        d,
      };
    };

    const options = functionalPage(C, 'test-page-name');
    expect(options.data).toMatchSnapshot();
    expect(fn).toBeCalledTimes(0);
    expect(fnOff).toBeCalledTimes(0);
    expect(fnLifeCycle).toBeCalledTimes(0);
  });

  test('throw if return illegal value', async () => {
    const c = function () {
      return function () {};
    };
    expect(() => {
      functionalPage(c, 'test-page-name');
    }).toThrowError();
  });

  interface IPageState {
    stateFoo: string;
    stateEffect: string;
    onShowData: boolean;
  }

  test('functional page 预渲染', async () => {
    const initStateValue = 'initValue';
    const callFn = vi.fn();
    const unloadFn = vi.fn();
    const unloadInOnLoad = vi.fn();

    const pageOption = functionalPage(function (props: any) {
      expect(props.hasOwnProperty('minifishHooks')).toBeFalsy();

      const [stateFoo] = useState(initStateValue);
      useEvent(
        'foo',
        (event) => {
          console.log('this is page handler');
        },
        [],
      );

      useEffect(() => {
        callFn();
        return unloadFn;
      }, []);

      useOnLoad((q) => {
        return unloadInOnLoad;
      }, []);

      useEvent('onTap', () => {}, []);
      return {
        stateFoo,
      };
    }, 'test-page-name');

    // 挂载前暂存数据
    expect(pageOption.data.stateFoo).toEqual(initStateValue);
    expect(pageOption.hasOwnProperty('foo')).toBeTruthy();
    expect(pageOption.hasOwnProperty('onTap')).toBeTruthy();
    expect(!pageOption.hasOwnProperty('methods')).toBeTruthy();
    expect(callFn).toBeCalledTimes(0);
    expect(unloadFn).toBeCalledTimes(0);
    expect(unloadInOnLoad).toBeCalledTimes(0);
  });

  test('functional page', async () => {
    const initStateValue = 'initValue';
    const valueAfterEffect = 'valueAfterEffect';
    const valueAfterLoad = 'valueAfterLoad';
    const valueAfterTap = 'valueAfterTap';
    const queryValue = 'queryValue';
    const callFn = vi.fn();
    const unloadFn = vi.fn();
    const unloadFn2 = vi.fn();
    const unloadFn3 = vi.fn();
    const unloadInOnLoad = vi.fn();
    const loadFn = vi.fn();

    const pageOption = functionalPage(function (props: any) {
      expect(props.hasOwnProperty('minifishHooks')).toBeFalsy();
      useEvent(
        'foo',
        (event) => {
          console.log('this is page handler');
        },
        [],
      );

      const [stateFoo, setFooState] = useState(initStateValue);
      const [stateEffect, setStateEffect] = useState(initStateValue);
      const [onShowData, setOnShowData] = useState(false);

      useEffect(() => {
        setStateEffect(valueAfterEffect);
        callFn();
        return unloadFn;
      }, []);

      useOnLoad((q) => {
        assert(q.query === queryValue);
        setFooState(valueAfterLoad);
        loadFn();

        return unloadInOnLoad;
      }, []);

      useOnShow(() => {
        setOnShowData(true);

        return unloadFn2;
      }, []);

      useEvent(
        'onTap',
        () => {
          if (onShowData) {
            setFooState(valueAfterTap);
          }
          return unloadFn3;
        },
        [onShowData],
      );

      return { stateFoo, stateEffect, onShowData };
    }, 'test-page-name');

    // 确保前面的事情都已经结束
    await delay(10);

    // 开始挂载页面
    let pageData: Partial<IPageState> = {};
    const mockAppxPage = async (opt) => {
      const mockInstance = {
        $id: 'aa',
        setData() {
          pageData = { ...pageData, ...arguments[0] };
        },
      };

      // 模拟触发 onLoad
      const task = opt.onLoad.call(mockInstance, { query: 'queryValue' });
      await task;
      await delay(20); // 实际生效是异步的，值越大，报错概率越低 （看运气，偶现挂掉）
      expect(callFn).toBeCalledTimes(1);
      expect(loadFn).toBeCalledTimes(1);
      expect(unloadInOnLoad).toBeCalledTimes(0);
      expect(pageData.stateFoo).toEqual(valueAfterLoad);

      opt.onShow.call(mockInstance);
      await delay(10);
      expect(pageData.onShowData).toEqual(true);

      expect(unloadInOnLoad).toBeCalledTimes(0);

      opt.onTap.call(mockInstance);
      await delay(20); // 实际生效是异步的
      expect(pageData.stateFoo).toEqual(valueAfterTap);

      opt.onUnload.call(mockInstance);
      await delay(10);

      expect(unloadFn).toBeCalledTimes(callFn.mock.calls.length);
      expect(unloadFn).toBeCalledTimes(1);
      expect(unloadFn2).toBeCalledTimes(1);
      expect(unloadFn3).toBeCalledTimes(1);
      expect(unloadInOnLoad).toBeCalledTimes(1);
    };

    await mockAppxPage(pageOption);

    // 重新拉起来，应该是个新的组件
    let pageData2: Partial<IPageState> = {};
    const mockAppxPage2 = async (opt) => {
      const mockInstance = {
        $id: '22',
        setData() {
          pageData2 = { ...pageData2, ...arguments[0] };
        },
      };
      opt.onLoad.call(mockInstance, { query: 'queryValue' });
      await delay(10);
    };

    await mockAppxPage2(pageOption);
    expect(pageData2.stateEffect).toEqual(valueAfterEffect);
    expect(pageData2.onShowData).toEqual(false);
  });

  test('minium functional page lifecycle', async () => {
    const unloadInOnLoad = vi.fn();
    const unloadFn2 = vi.fn();

    const pageOption = functionalPage(function (props) {
      const [, setCounter] = useState(0);
      const [onShowData, setOnShowData] = useState(false);

      useOnLoad((q) => {
        setCounter(1);
        return unloadInOnLoad;
      }, []);

      useOnShow(() => {
        setOnShowData(true);
        return unloadFn2;
      }, []);

      return { onShowData };
    }, 'test-page-name');
    expect(unloadInOnLoad).toBeCalledTimes(0);
    expect(pageOption.onShow).toBeDefined();

    // 开始挂载页面
    let pageData: Partial<IPageState> = {};
    const mockAppxPage = async (opt) => {
      const mockInstance = {
        $id: 'aa',
        setData() {
          pageData = { ...pageData, ...arguments[0] };
        },
      };

      // 模拟框架触发各种事件。注意框架是不会 await 这些调用的
      opt.onLoad.call(mockInstance, { query: 'queryValue' });
      opt.onShow.call(mockInstance);
      opt.onUnload.call(mockInstance);

      await delay(10);
      expect(unloadFn2).toBeCalledTimes(1);
      expect(unloadInOnLoad).toBeCalledTimes(1);
    };

    await mockAppxPage(pageOption);
    await delay(10);
  });

  test('onShareAppMessage 只允许监听一次', async () => {
    expect(() => {
      const opt = functionalPage(function (props) {
        useOnShareAppMessage(() => {}, []);
        useOnShareAppMessage(() => {}, []);

        return {};
      });
      mountAlipayPage(opt);
    }).toThrowError(/不能再新增/);
  });

  test('在返回值里提交数据', async () => {
    const pageOption = functionalPage(function (props) {
      const [s, setS] = useState('foo');
      const [h, setH] = useState('hhh');
      const value = `value-${s}`;

      const heavyValue = useMemo(() => {
        return `heavy-${h}`;
      }, [h]);

      useOnLoad(() => {
        setS('bar');
      }, []);

      useEvent(
        'onTap',
        () => {
          setH(`tap-heavy`);
        },
        [],
      );

      return {
        value,
        heavyValue,
      };
    }, 'return-page');
    expect(pageOption.data).toEqual({
      value: 'value-foo',
      heavyValue: 'heavy-hhh',
    });
    const page = mountAlipayPage(pageOption);

    await delay(20);
    expect(page.setDataRecord).toEqual([{ value: 'value-bar' }]);

    pageOption.onTap.call(page);
    await delay(20);
    expect(page.setDataRecord).toEqual([
      { value: 'value-bar' },
      { heavyValue: 'heavy-tap-heavy' },
    ]);
  });

  test('函数初始化，期望抛错', async () => {
    const C = function (props) {
      throw new Error('init error');
    };

    const pageName = 'test-page-name';

    expect(() => {
      functionalPage(C, pageName);
    }).toThrowError(new RegExp(pageName));
  });

  test('不传入 PageName，组件也能正常使用', async () => {
    expect(() => {
      functionalPage(function (props) {
        throw new Error('init error');
      });
    }).toThrowError(/出错/);
  });

  test('不用 useMemo，会触发不必要的更新', async () => {
    const pageOption = functionalPage<{ s: string }>(function (props) {
      const [s, setS] = useState('foo');
      const [bar, setBar] = useState('bar');
      const longList = [props.s, s];

      useEvent(
        'onTap',
        () => {
          setS('bar');
        },
        [],
      );

      useEvent(
        'onTap2',
        () => {
          setBar('tap2');
        },
        [],
      );

      return {
        list: longList,
        bar,
      };
    }, 'return-page');

    const page = mountAlipayPage(pageOption);
    await delay(20);
    pageOption.onTap.call(page);
    await delay(20);
    pageOption.onTap2.call(page);
    await delay(20);
    const initSet = { list: [undefined, 'foo'] };
    const secondSet = { list: [undefined, 'bar'] };
    expect(page.setDataRecord).toEqual([
      initSet,
      secondSet,
      { bar: 'tap2', ...secondSet },
    ]); // 最后一次只更新 barh
  });

  test('使用 useMemo 控制大数据的更新率', async () => {
    const pageOption = functionalPage<{ s: string }>(function (props) {
      const [s, setS] = useState('foo');
      const [bar, setBar] = useState('bar');
      const longList = useMemo(() => {
        return [props.s, s];
      }, [props.s, s]);

      useEvent(
        'onTap',
        () => {
          setS('bar');
        },
        [],
      );

      useEvent(
        'onTap2',
        () => {
          setBar('tap2');
        },
        [],
      );

      return {
        list: longList,
        bar,
      };
    }, 'return-page');

    const page = mountAlipayPage(pageOption);
    await delay(20);
    pageOption.onTap.call(page);
    await delay(20);
    pageOption.onTap2.call(page);
    await delay(20);
    const initSet = { list: [undefined, 'foo'] };
    const secondSet = { list: [undefined, 'bar'] };
    expect(page.setDataRecord).toEqual([initSet, secondSet, { bar: 'tap2' }]); // 最后一次只更新 barh
  });

  test('一些属性是不允许注册的', async () => {
    expect(() => {
      functionalPage(() => {
        useEvent('mixins', () => {}, []);
        return {};
      }, 'foo');
    }).toThrow(/不允许注册/);
  });

  test('支付宝端特有的生命周期', async () => {
    const C = function (props) {
      useOnTitleClick(() => {}, []);
      return {};
    };
    alipayPage(C);
    expect(() => {
      wechatPage(C);
    }).toThrow(/平台配置/);
  });
});

describe('component - wechat', async () => {
  let off: () => void;
  beforeEach(() => {
    off = setupWechatEnv();
  });

  afterEach(() => {
    off();
  });

  test('wechat page', async () => {
    const C = function () {
      const [s] = useState('foo');
      return {
        s,
      };
    };
    const instance = wechatPage(C);
    expect(instance.data).toEqual({ s: 'foo' });
  });
});

const wewe = () => {
  useEffect(() => {
    document.addEventListener('click', () => {});
  }, []);

  return {
    name: 'w',
  };
};
