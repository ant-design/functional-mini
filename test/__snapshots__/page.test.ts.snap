// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`page - common and alipay > minium get appx options 1`] = `
{
  "d": "d-initValue",
  "foo": "aa",
}
`;
